<a href="https://colab.research.google.com/github/sivanathvenigalla/Jaya-Venkatasivanath_INFO5731_Fall2024/blob/main/Venigalla_jayavenkatasivanathAssignment_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

In [17]:
from google.colab import files

uploaded = files.upload()

Saving Assignment1_denton_housing.csv to Assignment1_denton_housing.csv


This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [24]:
# Your code here
import math

def hemisphere_volume(radius):
    return (2 / 3) * math.pi * (radius ** 3)


print("r    V")

for r in range(2, 21, 2):
    volume = hemisphere_volume(r)

    print(f"{r:<4} {volume:.2f}")



r    V
2    16.76
4    134.04
6    452.39
8    1072.33
10   2094.40
12   3619.11
14   5747.02
16   8578.64
18   12214.51
20   16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [25]:

import random

random_num = [random.randint(50, 150) for _ in range(12)]
print("Random numbers:", random_num)

ascending_sorted_numbers = sorted(random_num)
print("Ascending order:", ascending_sorted_numbers)

descending_sorted_numbers = sorted(random_num ,reverse=True)
print("Descending order:", descending_sorted_numbers)

average_all = sum(random_num) / len(random_num)
print("Average of all numbers:", average_all)

average_first_6_ascending = sum(ascending_sorted_numbers[:6]) / 6
average_first_6_descending = sum(descending_sorted_numbers[:6]) / 6

# Result Display

print("Average of  first 6 numbers (ascending):", average_first_6_ascending)
print("Average of  first 6 numbers (descending):", average_first_6_descending)



Random numbers: [116, 116, 116, 147, 75, 51, 126, 54, 131, 115, 140, 106]
Ascending order: [51, 54, 75, 106, 115, 116, 116, 116, 126, 131, 140, 147]
Descending order: [147, 140, 131, 126, 116, 116, 116, 115, 106, 75, 54, 51]
Average of all numbers: 107.75
Average of  first 6 numbers (ascending): 86.16666666666667
Average of  first 6 numbers (descending): 129.33333333333334


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [26]:
# Your code here
import string
from collections import Counter

def sanitize_sentence(sentence):
    """Remove punctuation from the sentence."""
    return sentence.translate(str.maketrans('', '', string.punctuation))

def count_vowels(sentence):
    """Count the number of vowels in the sentence."""
    vowels = 'aeiou'
    counts = Counter(char for char in sentence.lower() if char in vowels)
    return {vowel: counts.get(vowel, 0) for vowel in vowels}

sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

words1 = sentence1.split()
words2 = sentence2.split()
print("List of words including punctuation:")
print(words1 + words2)

common_words = set(words1) & set(words2)
print("Words that occur in both sentences:", common_words)

sanitized_sentence1 = sanitize_sentence(sentence1)
sanitized_sentence2 = sanitize_sentence(sentence2)
print("Sanitized sentence 1:", sanitized_sentence1)
print("Sanitized sentence 2:", sanitized_sentence2)

vowel_counts1 = count_vowels(sanitized_sentence1)
vowel_counts2 = count_vowels(sanitized_sentence2)
print("Sentence 1 vowel counts:", ', '.join(f"{vowel}:{count}" for vowel, count in vowel_counts1.items()))
print("Sentence 2 vowel counts:", ', '.join(f"{vowel}:{count}" for vowel, count in vowel_counts2.items()))



Enter the first sentence: hello stupid 
Enter the second sentence: get lost
List of words including punctuation:
['hello', 'stupid', 'get', 'lost']
Words that occur in both sentences: set()
Sanitized sentence 1: hello stupid 
Sanitized sentence 2: get lost
Sentence 1 vowel counts: a:0, e:1, i:1, o:1, u:1
Sentence 2 vowel counts: a:0, e:1, i:0, o:1, u:0


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [28]:
# Your code here
import math

def evaluate_expression(expression):
    """___Evaluate the mathematical expression.___"""
    try:

        result = eval(expression, {"__builtins__": None}, {"sqrt": math.sqrt, "pow": pow, "mod": lambda x, y: x % y})
        return result
    except ZeroDivisionError:
        return "Error: Division by zero is not allowed."
    except (SyntaxError, NameError):
        return "Error: Invalid input expression."
    except Exception as e:
        return f"Error: {str(e)}"

def calculator():
    """Run the calculator program."""
    print("Simple Command-Line Calculator")
    print("Supported operations: +, -, *, /, ** (exponentiation), sqrt (square root), % (modulus)")
    print("Type 'exit' to quit the program.")

    while True:
        expression = input("Enter expression: ").strip()

        if expression.lower() == 'exit':
            print("Exiting the calculator.")
            break

        result = evaluate_expression(expression)
        print("Result:", result)

if __name__ == "__main__":
    calculator()

Simple Command-Line Calculator
Supported operations: +, -, *, /, ** (exponentiation), sqrt (square root), % (modulus)
Type 'exit' to quit the program.
Enter expression: 2+2
Result: 4
Enter expression: exit
Exiting the calculator.


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [30]:

import pandas as pd
from scipy.stats import hmean

filename = 'Assignment1_denton_housing.csv'
data = pd.read_csv(filename)

pivot_table = data.pivot(index='year', columns='title_field', values='value')

pivot_table = pivot_table.reset_index()


print("Pivoted DataFrame:")
print(pivot_table)

pivot_table['Occupied Housing Units Difference'] = pivot_table['Occupied Housing Units'].diff()
differences = pivot_table['Occupied Housing Units Difference'].dropna().astype(int)
sorted_differences = sorted(differences)
print("\nDifferences in Occupied Housing Units from year to year (ascending order):")
print(sorted_differences)

print("\nYear Vacant Housing Units Occupied Housing Units")
for index, row in pivot_table.iterrows():
    total_units = row['Total Housing Units']
    if total_units > 0:
        vacant_percentage = (row['Vacant Housing Units'] / total_units) * 100
        occupied_percentage = (row['Occupied Housing Units'] / total_units) * 100
        print(f"{row['year']} {vacant_percentage:.2f}% {occupied_percentage:.2f}%")
    else:
        print(f"{row['year']} ----- -----")

max_vacant_year = pivot_table.loc[pivot_table['Vacant Housing Units'].idxmax()]
max_occupied_year = pivot_table.loc[pivot_table['Occupied Housing Units'].idxmax()]

print(f"\nVacant Housing Units")
print(f"Year: {max_vacant_year['year']}, Value: {max_vacant_year['Vacant Housing Units']}")
print(f"Occupied Housing Units")
print(f"Year: {max_occupied_year['year']}, Value: {max_occupied_year['Occupied Housing Units']}")


non_zero_total_units = pivot_table['Total Housing Units'][pivot_table['Total Housing Units'] > 0]

if len(non_zero_total_units) > 0:
    harmonic_mean = hmean(non_zero_total_units)
    print(f"\nHarmonic Mean of Total Housing Units: {harmonic_mean:.2f}")
else:
    print("\nNo valid total housing units to calculate the harmonic mean.")

Pivoted DataFrame:
title_field  year  Median Gross Rent  Occupied Housing Units  Owner Occupied  \
0            2008                768                   36711           17340   
1            2009                769                   35916           18057   
2            2010                780                   41007           18867   
3            2011                821                   40704           18478   
4            2012                842                   42108           20068   
5            2013                887                   43673           20845   
6            2014                905                   46295           21427   

title_field  Renter Occupied  Total Housing Units  Vacant Housing Units  
0                      19371                41194                  4483  
1                      17859                39499                  3583  
2                      22140                44642                  3635  
3                      22226                

# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

The assignment had a number of difficulties, including managing CSV data, developing reliable error correction in calculators, and carrying out intricate string manipulations. The varied activities, which included data analysis and the creation of useful tools, were interesting and fulfilling despite these challenges. Although there was enough time allotted, the learning process may be enhanced by more resources, automated testing, intermediate milestones, and more explicit instructions.